purchase_receipt_progress/models/purchase_order.py

```diff
@@ -0,0 +1,89 @@
+    _description = "Purchase Order Received by Category"
+
+    order_id = fields.Many2one("purchase.order")
+    category_id = fields.Many2one("product.category")
+    product_qty = fields.Float()
+    qty_received = fields.Float()
+    to_receive = fields.Float()
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    purchased_quantity_total = fields.Float(
+        string="Total Quantity", compute="_compute_purchased_quantity_total"
+    )
+    delivery_status = fields.Float(compute="_compute_delivery_status")
+    received_by_category_ids = fields.One2many(
+        "purchase.order.received.by.category",
+        "order_id",
+        string="Received by Category",
+        compute="_compute_received_by_category_ids",
+    )
+
+    @api.depends("order_line.product_qty")
+    def _compute_purchased_quantity_total(self):
+        for order in self:
+            order.purchased_quantity_total = sum(order.mapped("order_line.product_qty"))
+
+    @api.depends("order_line.product_qty", "order_line.qty_received")
+    def _compute_delivery_status(self):
+        for order in self:
+            received_quantity = sum(order.mapped("order_line.qty_received"))
+            if order.purchased_quantity_total and received_quantity:
+                order.delivery_status = (
+                    received_quantity / order.purchased_quantity_total
+                )
+            else:
+                order.delivery_status = 0
+
+    @api.depends("order_line.product_qty", "order_line.qty_received")
+    def _compute_received_by_category_ids(self):
+        for order in self:
+            purchase_order_received_by_category_obj = self.env[
+                "purchase.order.received.by.category"
+            ]
+            order_id = (
+                order._origin.id if isinstance(order.id, models.NewId) else order.id
+            )
+            purchase_order_received_by_category_obj.search(
+                [("order_id", "=", order_id)]
+            ).unlink()
+            order.received_by_category_ids = False
+            received_by_category_ids = []
+            for category in order.mapped("order_line.product_id.categ_id"):
+                lines = order.order_line.filtered(
+                    lambda l: l.product_id.categ_id == category
+                )
+                if lines:
+                    values = {
+                        "order_id": order_id,
+                        "category_id": category.id,
+                        "product_qty": sum(lines.mapped("product_qty")),
+                        "qty_received": sum(lines.mapped("qty_received")),
+                        "to_receive": sum(lines.mapped("to_receive")),
+                    }
+                    received_by_category_id = (
+                        purchase_order_received_by_category_obj.create(values)
+                    )
+                    received_by_category_ids.append((4, received_by_category_id.id))
```

Comment:
Instead of creating the record and then adding the ID to the field, you can directly tell Odoo that you want a new record to be created when assigning it to the field:
```suggestion
                    received_by_category_ids.append((0, 0, values))
```
Odoo will see that the first command is `0`, will retrieve the values in third position of the tuple, and create the new record by itself when needed.